third_party/nvidia/hopper/lib/Transforms/WarpSpecialization/PingPong.cpp

```diff
@@ -73,202 +274,655 @@ void getNestedFor(Region *partition,
-    Value syncBarrier = builder.create<arith::ConstantIntOp>(
-        loc, idx == 0 ? PING_BARRIER : PONG_BARRIER, 32);
-    builder.create<ttng::NamedBarrierWaitOp>(loc, syncBarrier, numThreads);
-
-    Operation *insertBefore = moveBackward(endOfGemm, forOp);
-    builder.setInsertionPoint(insertBefore);
-    auto loc2 = endOfGemm->getLoc();
-    Value arriveBarrier = builder.create<arith::ConstantIntOp>(
-        loc2, idx == 0 ? PONG_BARRIER : PING_BARRIER, 32);
-    builder.create<ttng::NamedBarrierArriveOp>(loc2, arriveBarrier, numThreads);
   }
+  return defaultCapability;
 }
 
+/// doPingPongSync pass: Insert pingpong barriers to the IR
 void doPingPongSync(triton::FuncOp &funcOp, unsigned numWarpGroups,
                     int capability) {
-  // Insert sync points in ForOp for consumer warp groups. Enable this pass
-  // when number of consumer warp groups == 2.
-  if (numWarpGroups != 3)
-    return;
-
-  SmallVector<scf::ForOp> loops;
-  // Identify ForOps for consumer warp groups. Check partitions to find regions
-  // of gemms.
-  DenseMap<unsigned, SmallVector<Operation *>> loopDepthPartition0;
-  DenseMap<unsigned, SmallVector<Operation *>> loopDepthPartition1;
+  auto moduleOp = funcOp->getParentOfType<ModuleOp>();
+  capability = getCapability(moduleOp, capability);
   for (auto &block : funcOp.getBody().getBlocks()) {
     for (Operation &bodyOp : block.getOperations()) {
       Operation *op = &bodyOp;
       if (auto wsOp = dyn_cast<ttg::WarpSpecializeOp>(op)) {
-        handleWarpSpec(wsOp);
+        handleWarpSpec(wsOp, capability);
       }
     }
   }
 }
 
+static Operation *getSplitOp(Operation *op) {
+  for (Value operand : op->getOperands()) {
+    if (auto result = dyn_cast<OpResult>(operand)) {
+      if (isa<tt::SplitOp>(result.getOwner())) {
+        return result.getOwner();
+      }
+    }
+  }
+  return nullptr;
+}
+
+static std::optional<NameLoc> getNameLoc(Operation *op) {
+  Location loc = op->getLoc();
+  // Check if loc is directly a NameLoc
+  if (auto nameLoc = dyn_cast<NameLoc>(loc)) {
+    return nameLoc;
+  }
+  // Check if loc is a CallSiteLoc with a NameLoc callee
+  if (auto callSiteLoc = dyn_cast<CallSiteLoc>(loc)) {
+    if (auto nameLoc = dyn_cast<NameLoc>(callSiteLoc.getCallee())) {
+      return nameLoc;
+    }
+  }
+  // Check if loc is a FusedLoc and extract NameLoc from it
+  if (auto fusedLoc = dyn_cast<FusedLoc>(loc)) {
+    for (Location subLoc : fusedLoc.getLocations()) {
+      if (auto nameLoc = dyn_cast<NameLoc>(subLoc)) {
+        return nameLoc;
+      }
+      if (auto callSiteLoc = dyn_cast<CallSiteLoc>(subLoc)) {
+        if (auto nameLoc = dyn_cast<NameLoc>(callSiteLoc.getCallee())) {
+          return nameLoc;
+        }
+      }
+    }
+  }
+  return std::nullopt;
+}
+
+/// Check if two operations have equivalent locations
+/// Returns true if they have matching NameLoc or equivalent raw Location
+static bool areLocationsEquivalent(Operation *op1, Operation *op2) {
```

Comment:
This is suspicious, why do we need to look at the location of the original source?

Reply:
How would this work for something like TLX?